Quiz game

nkhaimov opened the pull request:
In this pull request, we added the following features:

- Verb conjugation quiz created from text by parsing sentences for verbs and creating answer choices with various conjugations.
- Navigation pane for changing between questions.
- Submit button grades quiz, shows score, and marks questions correct/incorrect with comparison of correct answer and user answer.
- Popup to confirm submit.
- General screen for selecting which text the quiz is generated from. (Right now the "Add Text" button does nothing)
![image](https://user-images.githubusercontent.com/55002917/106739817-96fdd380-65e7-11eb-83be-5e181a54f2e2.png)
![image](https://user-images.githubusercontent.com/55002917/106739626-5bfba000-65e7-11eb-8bb8-a55cb21f9dfd.png)
![image](https://user-images.githubusercontent.com/55002917/106739775-84839a00-65e7-11eb-9f73-fef28b492e12.png)

@ryaanahmed

montsegarza committed:
Added two conjugation functions from Nodebox and test cases

nkhaimov committed:
created quizView directory in frontend

nkhaimov committed:
Merge branch 'anagram' into quiz-game

nkhaimov committed:
added QuizView

nkhaimov committed:
Revert "added QuizView"

This reverts commit e0a78d2613a949a6873aaadb6b32628897b49d12.

nkhaimov committed:
Revert "Merge branch 'anagram' into quiz-game"

This reverts commit bfabdd414494364fd57fa851a98100476cd013c0, reversing
changes made to 3c3819e92dfda008c4cecc4b9e47ee6fad4125b4.

JusticeV452 committed:
Updated added get_quiz_data request

- Added funtion to create sentences (struggles with sentences with punctuations mid-sentence)
- Added function to create random quiz answer choices
- Added Text model and serializer
- Added api request
- Added __init__.py as verb_conjugation.py and verb.txt from nodebox_linguistics_extended

JusticeV452 committed:
Revert "Updated added get_quiz_data request"

This reverts commit 0134d59b846b47802eea3ade78f5f2f8cba7d183.

JusticeV452 committed:
Revert "Revert "Updated added get_quiz_data request""

This reverts commit e870f92e8a2cacc086afe1131651880aa209cef0.

JusticeV452 committed:
restore misc.xml

JusticeV452 committed:
Update misc.xml

JusticeV452 committed:
Update get_quiz_data endpoint

JusticeV452 committed:
Updated final sentence formatting

JusticeV452 committed:
Updated final sentence formatting

JusticeV452 committed:
Updated final sentence formatting

L-Jewel committed:
the quiz page appears and loads now

L-Jewel committed:
Fixed linter errors

JusticeV452 committed:
Changed get_quiz_sentences to also randomly select which verb to replace

nkhaimov committed:
styled quizView to match wireframe

JusticeV452 committed:
Implemented importing data from backend into QuizView

sharabhuiyan committed:
removed old verb conjugation functions

JusticeV452 committed:
Updated comments and doc string for get_quiz_sentences

JusticeV452 committed:
Made submit button move to the next question

JusticeV452 committed:
Merge branch 'quiz-game' of https://github.com/dhmit/lang_learn into quiz-game

sharabhuiyan committed:
updated get_quiz_sentences to remove contractions in input text

L-Jewel committed:
fixed linter errors

L-Jewel committed:
Added the dummy exit button for the quiz page

adanna-a committed:
Capitlize options in get_sentence_options and remove extra spaces in get_quiz_sentences

adanna-a committed:
Merge branch 'quiz-game' of https://github.com/dhmit/lang_learn into quiz-game

JusticeV452 committed:
Added test case for get_sentence_options

JusticeV452 committed:
Merge branch 'quiz-game' of https://github.com/dhmit/lang_learn into quiz-game

L-Jewel committed:
Tweaked top row formatting + added exit button

Forgot to mention this in the last commit, but I got the div to take up the whole page by making the height and width 100%

sharabhuiyan committed:
created function to process text

need to test on actual input from front end since python console can be weird about quotations

L-Jewel committed:
Reformatted the question box

TODO: Potentially reformat the buttons? I'm not very familiar with how these work, so I'm not sure how to do that.

L-Jewel committed:
Merge branch 'quiz-game' of https://github.com/dhmit/lang_learn into quiz-game

L-Jewel committed:
Added navigation buttons at the bottom

TODO: They're off-centered (in respect to the question box above) and I'm not sure how to fix that

L-Jewel committed:
Added navigational functionality

L-Jewel committed:
Added a scrolling feature on the left-hand-side that shows all of the questions

L-Jewel committed:
Progress bar allows you to click around questions

L-Jewel committed:
Added a much needed break between the progress bar elements

L-Jewel committed:
Fixed linter errors

L-Jewel committed:
Added a feature that keeps track of the users answers

In order to recieve data from the MC radio buttons, I had to integrate the ButtonChoices function into the QuizView class.

L-Jewel committed:
Reformatted progress bar + addressed issues with updating state

L-Jewel committed:
reformatted progress bar

added that nice aesthetic line on the left

JusticeV452 committed:
Preserve state of selected buttons

adanna-a committed:
Merge branch 'quiz-game' of https://github.com/dhmit/lang_learn into quiz-game

adanna-a committed:
Revert "Merge branch 'quiz-game' of https://github.com/dhmit/lang_learn into quiz-game"

This reverts commit 81e259ed6c6e56a574a29128f53f231b06f6e36f, reversing
changes made to 04c662552f07809eeed3beecf599415fea05b69c.

adanna-a committed:
Revert "Revert "Merge branch 'quiz-game' of https://github.com/dhmit/lang_learn into quiz-game""

This reverts commit 0914136b4512f3d9a9ac7a242daf5454652bc0a6.

nkhaimov committed:
updated arrows on prev/next buttons

nkhaimov committed:
created grading feature

nkhaimov committed:
added popup for submit button

sharabhuiyan committed:
fixed sentence and answer choice errors with single quotes

sentences with double quotes are still buggy, they do not appear as they should in the sentences we return

JusticeV452 committed:
Update requirements.txt

JusticeV452 committed:
Updated Tests

Added get_quiz_sentences test
Updated get_sentence_options test
Fixed potential index error for remove_contractions

adanna-a committed:
Merge branch 'quiz-game' of https://github.com/dhmit/lang_learn into quiz-game

sharabhuiyan committed:
added comments with some issues that need to be resolved

sharabhuiyan committed:
Merge branch 'quiz-game' of https://github.com/dhmit/lang_learn into quiz-game

JusticeV452 committed:
Updated get_sentence_options test

Updated the test to assert that all answer choices are different

nkhaimov committed:
centered nav buttons

L-Jewel committed:
Altered results page

L-Jewel committed:
Merge branch 'quiz-game' of https://github.com/dhmit/lang_learn into quiz-game

nkhaimov committed:
updated scrollbar

L-Jewel committed:
Progress bar is updated after grading

Instead of showing current/answered/unanswered, it shows correct/incorrect/unanswered.

L-Jewel committed:
Merge branch 'quiz-game' of https://github.com/dhmit/lang_learn into quiz-game

adanna-a committed:
add to conjugation.py

fix contractions, commas, and indexing for verb

adanna-a committed:
Merge branch 'quiz-game' of https://github.com/dhmit/lang_learn into quiz-game

L-Jewel committed:
Added a confirmation popup to the submit button

L-Jewel committed:
Added a confirm popup when the user hits submit

For some reason, I don't think the linter likes alerts?

L-Jewel commented:
I added a JS confirm popup so that the user could confirm submit before actually submitting the quiz. However, it's come to my attention that the linter might require us to develop a custom confirm submit window. Should we revert back to the hover feature (where hovering over the submit button would ask if you were sure), or keep it as is? @ryaanahmed

JusticeV452 committed:
Added general quiz page to list texts

JusticeV452 committed:
Merge branch 'master' into quiz-game

JusticeV452 committed:
Fix errors from update-from-master

JusticeV452 committed:
Removed conflicting styles and added Navbar and Footer

JusticeV452 committed:
Update index.js

JusticeV452 committed:
Added react-tooltip and readded AnagramView to app.js

JusticeV452 committed:
Fix errors from update-from-master

JusticeV452 committed:
Updated LoadingPage and added LoadingPage to quizView

Also fixed pylint errors and readded anagram tests

adanna-a committed:
Merge branch 'quiz-game' of https://github.com/dhmit/lang_learn into quiz-game

adanna-a committed:
Fixed spacing issues

- spacing with contractions and quotes

Co-Authored-By: sharabhuiyan <34693697+sharabhuiyan@users.noreply.github.com>

adanna-a committed:
Check for extra punctuation at the end

- fix errors that come with user punctuation mistakes
- fixed spacing for commas

Co-Authored-By: sharabhuiyan <34693697+sharabhuiyan@users.noreply.github.com>

ryaanahmed committed:
Merge branch 'master' into quiz-game

ryaanahmed committed:
remake initial migration

ryaanahmed committed:
lint

ryaanahmed commented on backend/app/quiz_creation/conjugation_quiz.py:
```diff
@@ -0,0 +1,189 @@
+    return options
+
+
+def process_text(text):
```
This name is vague ("process" could mean any data transformation at all). I think this work might be a bit redundant with some work over in `app.analysis.parts_of_speech` -- let's do a cleanup refactor so that all of this text cleanup stuff is in one place.

ryaanahmed commented on backend/app/quiz_creation/conjugation_quiz.py:
```diff
@@ -0,0 +1,189 @@
+
+        # Removes spaces with words, commas, and quotations
+        # "~~" is a filler to preserve length of list when indexing for the verb
+        if current_sentence['sentence'] != [] and (pos == ',' or pos == "POS" or word == "'" or
```
nitpick: easier to read if the breaks are aligned with the meaning of the code:

```
if (current_sentence['sentence'] != [] 
    and (pos == ',' or pos == "POS" or word == "'" or word == "''")):
```

ryaanahmed commented on backend/app/quiz_creation/conjugation_quiz.py:
```diff
@@ -0,0 +1,189 @@
+
+
+
+def get_quiz_sentences(text):
```
This function needs a refactor for clarity. Right now, it consumes tagged words left to right, does some processing on those words as it goes, and then when we find that the next token is a sentence-ending punctuation mark, we do a big inner loop of adding the right blank and the correct answer to the data structure.

While this works just fine, it means to understand the function, you have to read the whole thing linearly, and if something breaks in a particular part it'll be difficult to debug.

Here's my suggestion for a refactor: _first_ tokenize the whole text into sentences. Then per sentence run the inner loops from the existing function here, but broken out into helpers. It'll look something like this:

```
sentences = sentence_tokenize(text)  # NLTK has a sentence tokenizer that we could use
possible_quiz_sentence_dicts = []  # these might not pass validation bc they don't have answers

for sentence in sentences:
    current_sentence = {
        'sentence': [],
        'options': [],
        'answer': None
    }
    sentence_text = process_text(sentence)
    pos_tags = get_parts_of_speech_tags(sentence_text)
    for i, [word, pos] in enumerate(pos_tags):
        # factor this whole chunk out into a helper function
    possible_quiz_sentence_dicts.append(current_sentence)

# once we're done preprocessing the sentences, iterate over the possible sentences....
for sentence_dict in possible_quiz_sentence_dicts:
    # now the whole chunk of logic under
    #        if pos == '.' or i == len(pos_tags) - 1:
    # called out...
    select_verb_and_make_fillin(sentence_dict)
    if has_valid_answer(sentence_dict):
        quiz_sentences.append(current_sentence)
```

This way the purpose of each 20-30 line chunk of code is clearly named and factored out, can be _tested_ individually, and this big function will become a lot clearer.

ryaanahmed commented on backend/app/tests.py:
```diff
@@ -52,3 +54,59 @@ def test_get_anagrams(self):
         all_anagrams_2 = anagrams.get_anagrams(anagrams.get_letter_freq(word_2))
         self.assertTrue(some_anagrams_2.issubset(all_anagrams_2))
+
+    def test_get_sentence_options(self):
```
Great tests!

ryaanahmed commented on frontend/src/quizView/quizView.js:
```diff
@@ -0,0 +1,287 @@
+            );
+        });
+
+        return (
```
See my comment from review of #9 about splitting this up into some helper functions.

ryaanahmed committed:
update dependencies

ryaanahmed commented on requirements.txt:
```diff
@@ -1,54 +1,10 @@
-args==0.1.0
```
Ah looks like we had a lot of stuff in here from when we briefly added some other deps that needed cleanup. Thanks for doing that!

ryaanahmed reviewed (approved):
HUGE work. As with #9 merging this now and leaving my comments here for future refactor and cleanup: this is good enough for now, and nicely tested etc., but we need to do some refactor of the biggest functions here so that the project is maintainable.
